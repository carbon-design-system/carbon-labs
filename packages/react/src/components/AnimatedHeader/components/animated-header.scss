/**
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

@use 'AnimatedHeader/animated-header.scss';
@use 'HeaderTitle/header-title.scss';
@use 'Tiles/AITile/ai-tile.scss';
@use 'Tiles/AIPromptTile/ai-prompt-tile.scss';
@use 'Tiles/GlassTile/glass-tile.scss';
