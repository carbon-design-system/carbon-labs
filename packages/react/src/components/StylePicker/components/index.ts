/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export { StylePickerColor } from './style-picker-color';
export { StylePickerIcon } from './style-picker-icon';
export { StylePickerGroup } from './style-picker-group';
export { StylePickerSection } from './style-picker-section';
export { StylePickerSections } from './style-picker-sections';
export { StylePickerOption } from './style-picker-option';
export { StylePicker } from './style-picker';
