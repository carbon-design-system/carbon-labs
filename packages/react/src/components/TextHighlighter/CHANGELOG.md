# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2025-01-30)


### Features

* **TextHighlighter:** Add TextHighlighter component - React ([#384](https://github.com/carbon-design-system/carbon-labs/issues/384)) ([61dca6b](https://github.com/carbon-design-system/carbon-labs/commit/61dca6ba71aafcab4a101cb21fc898a630c4c333))
