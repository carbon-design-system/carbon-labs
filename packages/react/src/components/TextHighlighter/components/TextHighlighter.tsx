/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */
import React from 'react';
import { usePrefix } from '@carbon-labs/utilities/es/index.js';

/** Primary UI component for user interaction */

interface TextHighlighterProps {
  children?: React.ReactNode;
  kind: string;
  type: string;
  reference?: string;
}

export const TextHighlighter = ({
  children,
  kind = 'mark',
  type = 'default',
  reference,
  ...rest
}: TextHighlighterProps) => {
  const prefix = usePrefix();
  let containerClass = '';
  let BaseTag: React.ElementType = 'mark';
  switch (kind) {
    case 'ins':
      containerClass = `${prefix}--text-highlighter__container ${prefix}--text-highlighter__container__ins`;
      BaseTag = 'ins';
      break;
    case 'del':
      containerClass = `${prefix}--text-highlighter__container ${prefix}--text-highlighter__container__del`;
      BaseTag = 'del';
      break;
    default:
      containerClass = `${prefix}--text-highlighter__container ${prefix}--text-highlighter-${type}`;
  }

  return (
    <div className={containerClass}>
      <BaseTag>{children}</BaseTag>
      {kind === 'mark' && reference && <sup>{reference}</sup>}
    </div>
  );
};
