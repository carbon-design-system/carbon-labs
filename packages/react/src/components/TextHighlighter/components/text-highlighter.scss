/**
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// Carbon setting imports
@use '@carbon/styles/scss/theme' as *;
@use '@carbon/styles/scss/colors' as *;
@use '@carbon/styles/scss/spacing' as *;

@use '@carbon/react/scss/components/tag/tokens' as tag;

$prefix: 'clabs' !default;

@mixin th-theme($bg-color, $text-color) {
  background-color: $bg-color;
  color: $text-color;
}

// -- Flushing the default styles for mark, del and ins --
mark {
  background-color: inherit;
  color: inherit;
}

del,
ins {
  color: inherit;
  text-decoration: none;
}
// -------------------------------------------------------

.#{$prefix}--text-highlighter__container {
  display: inline;
  padding: 0 $spacing-01;
  background-color: $yellow-20;
  color: $black;

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }

  sup {
    display: inline-block;
    font-size: 60%;
    margin-block-start: -$spacing-01 + 0.2;
    margin-inline-start: $spacing-02;
    vertical-align: top;
  }
}

.#{$prefix}--text-highlighter__container__ins {
  background-color: tag.$tag-background-green;
  border-block-end: 1px dashed;
  color: tag.$tag-color-green;

  @media screen and (-ms-high-contrast: active) {
    @include th-theme($background-inverse, $text-inverse);
  }
}

.#{$prefix}--text-highlighter__container__del {
  background-color: tag.$tag-background-red;
  color: tag.$tag-color-red;
  text-decoration: line-through;

  @media screen and (-ms-high-contrast: active) {
    @include th-theme($background-inverse, $text-inverse);
  }
}

// -- Color themes for text-heighlighter (mark) --

.#{$prefix}--text-highlighter-magenta {
  @include th-theme(tag.$tag-background-magenta, tag.$tag-color-magenta);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-purple {
  @include th-theme(tag.$tag-background-purple, tag.$tag-color-purple);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-blue {
  @include th-theme(tag.$tag-background-blue, tag.$tag-color-blue);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-cyan {
  @include th-theme(tag.$tag-background-cyan, tag.$tag-color-cyan);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-teal {
  @include th-theme(tag.$tag-background-teal, tag.$tag-color-teal);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-gray {
  @include th-theme(tag.$tag-background-gray, tag.$tag-color-gray);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-cool-gray {
  @include th-theme(tag.$tag-background-cool-gray, tag.$tag-color-cool-gray);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-warm-gray {
  @include th-theme(tag.$tag-background-warm-gray, tag.$tag-color-warm-gray);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
.#{$prefix}--text-highlighter-high-contrast {
  @include th-theme($background-inverse, $text-inverse);

  @media screen and (-ms-high-contrast: active) {
    background-color: $yellow-30;
  }
}
// -------------------------------------------------------

// -- For screen reader --
mark::before,
mark::after,
del::before,
del::after,
ins::before,
ins::after {
  position: absolute;
  overflow: hidden;
  block-size: 1px;
  clip: rect(1px, 1px, 1px, 1px);
  clip-path: inset(100%);
  inline-size: 1px;
  white-space: nowrap;
}

mark::before {
  content: ' [highlight start] ';
}

mark::after {
  content: ' [highlight end] ';
}

del::before {
  content: ' [deletion start] ';
}

del::after {
  content: ' [deletion end] ';
}

ins::before {
  content: ' [insertion start] ';
}

ins::after {
  content: ' [insertion end] ';
}
