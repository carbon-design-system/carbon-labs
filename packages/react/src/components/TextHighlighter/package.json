{
  "name": "@carbon-labs/react-text-highlighter",
  "version": "0.1.0",
  "publishConfig": {
    "access": "public",
    "provenance": true
  },
  "type": "module",
  "description": "Carbon Labs - Text Highlighter",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/carbon-design-system/carbon-labs",
    "directory": "src/components/TextHighlighter"
  },
  "main": "./src/index.js",
  "module": "./src/index.js",
  "files": [
    "es",
    "lib",
    "scss"
  ],
  "scripts": {
    "build": "node ../../../tasks/build.js",
    "clean": "rm -rf {es,lib,scss}"
  },
  "devDependencies": {
    "@carbon-labs/utilities": "canary"
  }
}
