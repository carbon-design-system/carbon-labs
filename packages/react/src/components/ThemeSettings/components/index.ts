/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

export { ThemeSettings } from './ThemeSettings';
export { ThemeSwitcher } from './ThemeSwitcher';
export { ThemeSetDropdown } from './ThemeSetDropdown';
export { ThemeMenuComplement } from './ThemeMenuComplement';
export { themeSets } from './theme-settings-consts';
