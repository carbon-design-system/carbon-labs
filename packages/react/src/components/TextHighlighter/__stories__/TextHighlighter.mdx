import { ArgTypes, Canvas, Meta, Controls } from '@storybook/blocks';
import * as TextHighlighterStories from './TextHighlighter.stories';

<Meta isTemplate />

# TextHighlighter

- **Initiative owner(s):** Punnoose Wilson
- **Status:** Draft
- **Target library:** TBD
- **Target library maintainer(s) / PR Reviewer(s):** N/A
- **Support channel:** `#carbon-labs`

{/* <!-- START doctoc generated TOC please keep comment here to allow auto update --> */}
{/* <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE --> */}

> ðŸ’¡ Check our
> [Stackblitz](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/TextHighlighter)
> example implementation.

[![Edit carbon-labs](https://developer.stackblitz.com/img/open_in_stackblitz.svg)](https://stackblitz.com/github/carbon-design-system/carbon-labs/tree/main/examples/react/TextHighlighter)

## Table of Contents

- [Overview](#overview)
  - [Variants](#here-we-have-a-component-offering-3-variants)
- [Getting started](#getting-started)
- [Example usages](#example-usages)
  - [Version comparison](#1-version-comparison)
  - [Reference anotation](#2-reference-anotation)
  - [Document versioning](#3-document-versioning)
  - [Document highlighting with source](#4-document-highlighting-with-source)

{/* <!-- END doctoc generated TOC please keep comment here to allow auto update --> */}

## Overview

The text highlighter feature enhances user experience by improving readability, focus, and engagement with the content. This feature is especially beneficial in various contexts such as version comparison, commenting, documentation etc

***Itâ€™s crucial that we use HTML semantics appropriately to represent the text highlighters, as improper usage could lead to a negative user experience. We must ensure effective communication to users utilizing assistive technologies, necessitating careful control over the markup.***

### Here we have a component offering 3 variants:

- 1st variant - `<mark>` : This HTML element represents text which is marked or highlighted for reference or notation purposes.
- 2nd variant - `<ins>` : This HTML element represents a range of text that has been added to a document. (Comparison purpose)
- 3rd variant - `<del>` : This HTML element represents a range of text that has been deleted from a document. (Comparison purpose)

<Canvas of={TextHighlighterStories.Default} />

<Controls />

## Getting started

Here's a quick example to get you started.

```bash
yarn add @carbon/react
yarn add @carbon-labs/react-text-highlighter
```

### JS (via import)

```javascript
import { TextHighlighter } from '@carbon-labs/react-text-highlighter/es/index';

function App() {
  return <TextHighlighter />;
}

export default App;
```

### SCSS

In your styles file import

```
@use '@carbon/react';
@use '@carbon-labs/react-text-highlighter/scss/text-highlighter';
```

## Example usages

Text highlighters add significant value by making content more interactive, engaging, and easier to manage and understand.

### **1. Version comparison**

<Canvas of={TextHighlighterStories.VersionComparison} />

### **2. Reference anotation**

<Canvas of={TextHighlighterStories.ReferenceAnotation} />

### **3. Document versioning**

<Canvas of={TextHighlighterStories.DocumentVersioning} />

### **4. Document highlighting with source**

<Canvas of={TextHighlighterStories.DocumentHighlightingWithSource} />
