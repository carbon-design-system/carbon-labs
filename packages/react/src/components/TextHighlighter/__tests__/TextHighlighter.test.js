/**
 * @license
 *
 * Copyright IBM Corp. 2025
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { render } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';

import { TextHighlighter } from '../components/TextHighlighter';
jest.mock('./text-highlighter.scss', () => ({}));
describe('TextHighlighter', () => {
  describe('renders as expected - Component API', () => {
    it('should match snapshot', () => {
      const { container } = render(<TextHighlighter />);
      expect(container).toMatchSnapshot();
    });
  });
});
