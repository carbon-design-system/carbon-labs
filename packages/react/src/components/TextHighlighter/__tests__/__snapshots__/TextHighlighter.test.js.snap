// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextHighlighter renders as expected - Component API should match snapshot 1`] = `
<div>
  <div
    class="clabs--text-highlighter__container clabs--text-highlighter-default"
  >
    <mark />
  </div>
</div>
`;
